Habitat should not expand $PATH at buildtime

james-stocks opened the pull request:
Habitat should have everything it needs at runtime as a habitat dependency and should prefer habitat dependency bins over native bins.

Signed-off-by: James Stocks <jstocks@chef.io>

Aha! Link: https://chef.aha.io/features/SH-2628

Commit:
Fix PATH for Habitat execution

plan.sh was expanding $PATH at build time when writing out the inspec bin script. This is incorrect, the variable should be expanded at runtime.

Signed-off-by: James Stocks <jstocks@chef.io>

stevendanna commented on habitat/plan.sh:
```diff
@@ -69,9 +69,6 @@ wrap_inspec_bin() {
 set -e
 
-# Set binary path that allows InSpec to use non-Hab pkg binaries
-export PATH="/sbin:/usr/sbin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:$PATH"
```
Another point of interest is that this `$PATH` will actually be evaluated at build-time and will include the build desp rather than just the runtime deps.

james-stocks replied:
I would have assumed this only affected build-time but I tested building an artifact with and without this line and it does change the `$PATH` when I'm running `hab pkg exec`

stevendanna replied:
Yes, sorry, what I mean is that `PATH` here is interpreted at build-time not run time. which is why you are getting all of the Hab binaries in your path.

stevendanna replied:
As in, this script this is rendering is called at runtime, but the script that is rendered doesn't contain `$PATH` it contains the build-time expansion of that environment variable.

james-stocks replied:
_puts on glasses_

Oh this is rendering the bin script for InSpec. And `$PATH` gets expanded at build time. So this was extra buggy.

james-stocks replied:
@stevendanna reviewing this with @TheLunaticScripter , we currently need to put native bin dirs first. But we need to fix the $PATH containing build deps. I changed the commit to simply escape the $PATH variable at build time so it expands at runtime.

stevendanna reviewed (commented):
I wonder if instead of completely removing this, we want to setup a PATH that would be similar to what `hab pkg exec` builds such that people who are using this package via binstubs get consistent behavior.  Namely, as of hab 1.5.96+, `hab pkg exec` will construct a PATH consisting of the paths for the package's runtime dep tree plus the incoming using environment's PATH.

The binstubs, however, are simple links into the package and do not get a hab controlled path.

We could potentially create a more consistent experience in our wrapper by doing something like (warning, code written in comment box):

```shell
USER_PATH=\$PATH
. ${pkg_prefix}/RUNTIME_ENVIRONMENT
export PATH=\$PATH:\$USER_PATH
```

One note that we may want to put in the release notes or something is that this may break people who are using `hab pkg exec` to execute hab on versions older than 1.5.96 since then they will not have access to system commands and since`inspec` has a number of resources that require a dependencies not currently in our `pkg_deps`.  (I believe this was done to avoid `inspec` depending a very large dependency tree of hab packages. For example the postgres_session resource depends on `psql`: https://github.com/inspec/inspec/blob/master/lib/inspec/resources/postgres_session.rb#L67)

Commit:
Remove world readable attribute from /usr/local/bundle

Artifact tests may fail due to warnings about `Insecure world writable dir /usr/local/bundle in PATH` appearing in stderr

Signed-off-by: James Stocks <jstocks@chef.io>
